packages/grafana-data/src/types/logs.ts

```diff
@@ -171,8 +173,25 @@ export const hasLogsContextSupport = (datasource: unknown): datasource is DataSo
   if (!datasource) {
     return false;
   }
-
+  // Typescript documentation recommends to use this approach to check if a property exists on an object
+  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
   const withLogsSupport = datasource as DataSourceWithLogsContextSupport;
 
   return withLogsSupport.getLogRowContext !== undefined && withLogsSupport.showContextToggle !== undefined;
 };
+
+/**
+ * Data sources that support log volume.
+ * This will enable full-range log volume histogram in Explore.
+ */
+export interface DataSourceWithLogsVolumeSupport<TQuery extends DataQuery> {
+  getLogsVolumeDataProvider(request: DataQueryRequest<TQuery>): Observable<DataQueryResponse> | undefined;
+}
+
+export const hasLogsVolumeSupport = <TQuery extends DataQuery>(
+  datasource: unknown
+): datasource is DataSourceWithLogsVolumeSupport<TQuery> => {
+  // Typescript documentation recommends to use this approach to check if a property exists on an object
+  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
```

Comment:
We could also do this without assertion, but assertion is recommended in Typescript docs: https://www.typescriptlang.org/docs/handbook/advanced-types.html#using-type-predicates
```
export const hasLogsVolumeSupport = <TQuery extends DataQuery>(
  datasource: unknown
): datasource is DataSourceWithLogsVolumeSupport<TQuery> => {
  if (!datasource || typeof datasource !== 'object') {
    return false;
  }
  const getLogsVolumeDataProviderGuard = (o: object): o is { getLogsVolumeDataProvider: unknown } =>
    'getLogsVolumeDataProvider' in o;

  return getLogsVolumeDataProviderGuard(datasource) && typeof datasource.getLogsVolumeDataProvider !== undefined;
};
```

Reply:
@joshhunt any (pun intended 😅) thoughts on what is preferred here?

Reply:
Preferably, we would upgrade to Typescript 9.4 and use the [improved `in` narrowing](https://devblogs.microsoft.com/typescript/announcing-typescript-4-9/#in-narrowing).

Otherwise, I think it's okay to type the incomming argument as `any` and just be short and simple about checking it in this function :)

I think that would be an okay exception to save to the betterer.results, which we can remove once we upgrade typescript.